hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java

```diff
@@ -154,25 +157,41 @@ public boolean recoverLease(Path f) throws IOException {
     }
 
     // finalize the final block and get block length
-    List<OmKeyLocationInfo> locationInfoList = keyInfo.getLatestVersionLocations().getLocationList();
+    List<OmKeyLocationInfo> locationInfoList = leaseKeyInfo.getKeyInfo().getLatestVersionLocations().getLocationList();
     if (!locationInfoList.isEmpty()) {
       OmKeyLocationInfo block = locationInfoList.get(locationInfoList.size() - 1);
       try {
         block.setLength(getAdapter().finalizeBlock(block));
       } catch (Throwable e) {
-        if (!forceRecovery) {
+        if (e instanceof StorageContainerException && (((StorageContainerException) e).getResult().equals(NO_SUCH_BLOCK)
+            || ((StorageContainerException) e).getResult().equals(CONTAINER_NOT_FOUND))
+            && !leaseKeyInfo.getIsKeyInfo() && locationInfoList.size() > 1) {
+          locationInfoList = leaseKeyInfo.getKeyInfo().getLatestVersionLocations().getLocationList().subList(0,
+              locationInfoList.size() - 1);
+          block = locationInfoList.get(locationInfoList.size() - 1);
```

Comment:
is this the same as
```suggestion
          block = locationInfoList.get(locationInfoList.size() - 2);
```

Reply:
This is same but we also use locationInfoList completely during [commit](https://github.com/apache/ozone/blob/c68afdc49e7c4ae44a64aca0a8c48235447a26ab/hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java#L191) so we want to truncate the last element.

Reply:
Yes code is duplicated across FileSystem classes, I will file a new Jira to remove these duplication.

Reply:
uh ok i see.